src/Symfony/Component/DependencyInjection/Compiler/AliasDeprecatedPublicServicesPass.php

```diff
@@ -0,0 +1,73 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Exception\InvalidArgumentException;
+use Symfony\Component\DependencyInjection\Exception\RuntimeException;
+use Symfony\Component\DependencyInjection\Reference;
+
+final class AliasDeprecatedPublicServicesPass extends AbstractRecursivePass
+{
+    private $tagName;
+
+    private $aliases = [];
+
+    public function __construct(string $tagName = 'deprecated_public_service')
+    {
+        $this->tagName = $tagName;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function processValue($value, bool $isRoot = false)
+    {
+        if ($value instanceof Reference && isset($this->aliases[$id = (string) $value])) {
```

Comment:
I just realize I'm actually chaging all the refences here contrary to what I wrote in the PR description. I need to recheck this part.

Reply:
I have checked and the behavior here is correct for me. We convert all references to the public service to the alias target service. This is only useful if the container is not dumped since this is already done when dumping normally.